src/data_structures/interval_tree.rs

```diff
@@ -0,0 +1,315 @@
+extern crate num;
+
+use self::num::traits::Num;
+
+use std::cmp;
+use std::mem;
+use std::fmt::Debug;
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct IntervalTree<N, D> {
+    root: Option<Node<N, D>>,
+}
+
+impl<N: Debug + Num + Clone + Ord, D: Debug> IntervalTree<N, D> {
+    pub fn new() -> Self {
+        IntervalTree { root: None }
+    }
+
+    pub fn insert(&mut self, start: N, end: N, data: D) {
+        validate_interval(&start, &end);
+        match self.root {
+            Some(ref mut n) => n.insert(start, end, data),
+            None => self.root = Some(Node::new(start, end, data)),
+        }
+    }
+
+    pub fn find_intersection(&self, start: N, end: N) -> Option<&D> {
```

Comment:
What do you think about returning a reference to the node here? This would allow to calculate the overlap (with whatever metric comes to the users mind). Also, what happens if there are multiple intersecting intervals? Shouln't this return an Iterator so people can choose whether they want exhaustive search or the first?